// @mainframe/rpc-error/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.isServerError = isServerError;
exports.getErrorMessage = getErrorMessage;
exports.internalError = exports.invalidParams = exports.methodNotFound = exports.invalidRequest = exports.parseError = exports.default = exports.ERROR_MESSAGE = exports.ERROR_CODE = void 0;

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _wrapNativeSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/wrapNativeSuper"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ERROR_MESSAGE;

var ERROR_CODE;
exports.ERROR_CODE = ERROR_CODE;

(function (ERROR_CODE) {
  ERROR_CODE[ERROR_CODE["PARSE_ERROR"] = -32700] = "PARSE_ERROR";
  ERROR_CODE[ERROR_CODE["INVALID_REQUEST"] = -32600] = "INVALID_REQUEST";
  ERROR_CODE[ERROR_CODE["METHOD_NOT_FOUND"] = -32601] = "METHOD_NOT_FOUND";
  ERROR_CODE[ERROR_CODE["INVALID_PARAMS"] = -32602] = "INVALID_PARAMS";
  ERROR_CODE[ERROR_CODE["INTERNAL_ERROR"] = -32603] = "INTERNAL_ERROR";
})(ERROR_CODE || (exports.ERROR_CODE = ERROR_CODE = {}));

var ERROR_MESSAGE = (_ERROR_MESSAGE = {}, _ERROR_MESSAGE[ERROR_CODE.PARSE_ERROR] = 'Parse error', _ERROR_MESSAGE[ERROR_CODE.INVALID_REQUEST] = 'Invalid request', _ERROR_MESSAGE[ERROR_CODE.METHOD_NOT_FOUND] = 'Method not found', _ERROR_MESSAGE[ERROR_CODE.INVALID_PARAMS] = 'Invalid params', _ERROR_MESSAGE[ERROR_CODE.INTERNAL_ERROR] = 'Internal error', _ERROR_MESSAGE);
exports.ERROR_MESSAGE = ERROR_MESSAGE;

function isServerError(code) {
  return -32000 >= code && code >= -32099;
}

function getErrorMessage(code) {
  return ERROR_MESSAGE[code] || (isServerError(code) ? 'Server error' : 'Application error');
}

var RPCError =
/*#__PURE__*/
function (_Error) {
  (0, _inheritsLoose2.default)(RPCError, _Error);

  RPCError.fromObject = function fromObject(err) {
    return new RPCError(err.code, err.message, err.data);
  };

  function RPCError(code, message, data) {
    var _this;

    _this = _Error.call(this) || this;
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "code", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "data", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "message", void 0);
    _this.code = code;
    _this.data = data;
    _this.message = message || getErrorMessage(code);
    return _this;
  }

  var _proto = RPCError.prototype;

  _proto.toObject = function toObject() {
    return {
      code: this.code,
      data: this.data,
      message: this.message
    };
  };

  return RPCError;
}((0, _wrapNativeSuper2.default)(Error));

exports.default = RPCError;

function createErrorFactory(code) {
  var message = ERROR_MESSAGE[code];
  return function createError(data) {
    return new RPCError(code, message, data);
  };
}

var parseError = createErrorFactory(ERROR_CODE.PARSE_ERROR);
exports.parseError = parseError;
var invalidRequest = createErrorFactory(ERROR_CODE.INVALID_REQUEST);
exports.invalidRequest = invalidRequest;
var methodNotFound = createErrorFactory(ERROR_CODE.METHOD_NOT_FOUND);
exports.methodNotFound = methodNotFound;
var invalidParams = createErrorFactory(ERROR_CODE.INVALID_PARAMS);
exports.invalidParams = invalidParams;
var internalError = createErrorFactory(ERROR_CODE.INTERNAL_ERROR);
exports.internalError = internalError;
