// @mainframe/rpc-request/esm/index.js
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import BaseRPC from '@mainframe/rpc-base';
import RPCError from '@mainframe/rpc-error';

var RequestRPC =
/*#__PURE__*/
function (_BaseRPC) {
  _inheritsLoose(RequestRPC, _BaseRPC);

  function RequestRPC(fetch) {
    var _this;

    _this = _BaseRPC.call(this, false) || this;

    _defineProperty(_assertThisInitialized(_this), "_fetch", void 0);

    _this._fetch = fetch;
    return _this;
  }

  var _proto = RequestRPC.prototype;

  _proto.request = function request(method, params) {
    return this._fetch({
      id: this.createId(),
      jsonrpc: '2.0',
      method: method,
      params: params
    }).then(function (msg) {
      if (msg.error) {
        throw RPCError.fromObject(msg.error);
      }

      return msg.result;
    });
  };

  return RequestRPC;
}(BaseRPC);

export { RequestRPC as default };
