// @mainframe/rpc-request/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _rpcBase = _interopRequireDefault(require("@mainframe/rpc-base"));

var _rpcError = _interopRequireDefault(require("@mainframe/rpc-error"));

var RequestRPC =
/*#__PURE__*/
function (_BaseRPC) {
  (0, _inheritsLoose2.default)(RequestRPC, _BaseRPC);

  function RequestRPC(fetch) {
    var _this;

    _this = _BaseRPC.call(this, false) || this;
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_fetch", void 0);
    _this._fetch = fetch;
    return _this;
  }

  var _proto = RequestRPC.prototype;

  _proto.request = function request(method, params) {
    return this._fetch({
      id: this.createId(),
      jsonrpc: '2.0',
      method: method,
      params: params
    }).then(function (msg) {
      if (msg.error) {
        throw _rpcError.default.fromObject(msg.error);
      }

      return msg.result;
    });
  };

  return RequestRPC;
}(_rpcBase.default);

exports.default = RequestRPC;
