// @mainframe/transport-http-node/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.resOrError = resOrError;
exports.default = createTransport;
exports.HTTPError = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

const HEADERS = {
  Accept: 'application/json',
  'Content-Type': 'application/json'
};
const METHOD = 'POST';

class HTTPError extends Error {
  constructor(status, message) {
    super(message);
    (0, _defineProperty2.default)(this, "status", void 0);
    this.status = status;
  }

}

exports.HTTPError = HTTPError;

function resOrError(res) {
  return res.ok ? Promise.resolve(res) : Promise.reject(new HTTPError(res.status, res.statusText));
}

function createTransport(url) {
  return function request(data) {
    const request = {
      body: JSON.stringify(data),
      headers: HEADERS,
      method: METHOD
    };
    return (0, _nodeFetch.default)(url, request).then(resOrError).then(res => res.json());
  };
}
